File: elixir/test/support/spike/agents.ex
defmodule Toolnexus.Spike.Agents do
  @moduledoc """
  SPIKE — Level-1/2 UX sugar over the substrate (port of js/spike/agents.ts).
  One new noun: `agent/2`. Everything compiles to the six verbs in Runtime.
  """

  alias Toolnexus.Spike.Runtime
  alias Toolnexus.ToolResult

  @bootstrap_order ["AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md"]
  @max_bootstrap_file_bytes 2 * 1024 * 1024
  @heartbeat_ok "HEARTBEAT_OK"

  def bootstrap_order, do: @bootstrap_order
  def heartbeat_ok, do: @heartbeat_ok

  @doc "The one new noun."
  def agent(name, spec), do: %{name: name, spec: Map.new(spec)}

  @doc "Collect this agent + its whole team graph into a runtime registry."
  def registry(a, acc \\ %{}) do
    if Map.has_key?(acc, a.name) do
      acc
    else
      spec = a.spec

      soul =
        cond do
          spec[:soul_file] -> File.read!(spec[:soul_file])
          spec[:soul] -> spec[:soul]
          true -> ""
        end

      team = spec[:team] || []

      acc =
        Map.put(acc, a.name, %{
          name: a.name,
          description: spec[:does],
          system_prompt: soul,
          model: spec[:model] || "inherit",
          tools: get_in(spec, [:uses, :tools]),
          budget: spec[:budget],
          wait_for: spec[:wait_for],
          on_spawn: spec[:on_spawn],
          on_close: spec[:on_close],
          # team scoping: task targets = ONLY this agent's team (not the whole registry)
          task_targets: Enum.map(team, & &1.name)
        })

      Enum.reduce(team, acc, fn t, acc -> registry(t, acc) end)
    end
  end

  @doc "Level 1: one-shot — build a runtime, run to completion, tear down."
  def run(a, rt_opts, prompt) do
    rt = Runtime.new(Map.put(Map.new(rt_opts), :registry, registry(a)))

    case Runtime.spawn_agent(rt, rt.root, a.name) do
      {:error, msg} ->
        %{text: msg, is_error: true, status: "done", pending: nil, turns: 0, total_tokens: 0, rt: rt}

      {:ok, h} ->
        Runtime.wake(rt, h, prompt)
        r = Runtime.wait(rt, h)
        if r.status != "pending", do: Runtime.close(rt)
        # exposed for durable resume
        Map.put(r, :rt, rt)
    end
  end

  @doc "Bridge: an Agent IS a Tool — drop it into the OLD API's extra tools."
  def as_tool(a, rt_opts) do
    Toolnexus.define_tool(%{
      name: a.name,
      description: a.spec[:does],
      input_schema: %{
        "type" => "object",
        "properties" => %{"prompt" => %{"type" => "string"}},
        "required" => ["prompt"]
      },
      execute: fn args ->
        r = run(a, rt_opts, to_string(args["prompt"] || ""))

        %ToolResult{
          output: r.text,
          is_error: r.is_error,
          metadata: %{agent: a.name, turns: r.turns, total_tokens: r.total_tokens}
        }
      end
    })
  end

  @doc """
  Level 2: the directory IS the agent. Discovers (all optional, openclaw order):
  AGENTS.md, SOUL.md, IDENTITY.md, USER.md, TOOLS.md, HEARTBEAT.md, MEMORY.md —
  injected as `## <file>` sections, capped at 2MB per file.
  """
  def agent_from_dir(dir, opts \\ %{}) do
    opts = Map.new(opts)

    sections =
      for f <- @bootstrap_order, path = Path.join(dir, f), File.exists?(path) do
        body = File.read!(path)

        body =
          if byte_size(body) > @max_bootstrap_file_bytes,
            do: binary_part(body, 0, @max_bootstrap_file_bytes) <> "\n[truncated: file exceeds bootstrap cap]",
            else: body

        "## #{f}\n\n#{String.trim(body)}"
      end

    name = opts[:name] || Path.basename(dir)

    spec =
      Map.merge(
        %{does: "persona agent from #{dir}", soul: Enum.join(sections, "\n\n")},
        Map.drop(opts, [:name])
      )

    agent(name, spec)
  end

  @doc """
  Level 2: start a long-lived persona — the heartbeat posts a tick to its own
  inbox (unsolicited rail; ticks coalesce) and wakes it; HEARTBEAT_OK results
  stay silent.
  """
  def start_agent(a, rt_opts, opts \\ %{}) do
    opts = Map.new(opts)
    rt = Runtime.new(Map.put(Map.new(rt_opts), :registry, registry(a)))
    {:ok, h} = Runtime.spawn_agent(rt, rt.root, a.name)
    on_report = opts[:on_report]

    looper =
      if opts[:every_ms] do
        every = opts[:every_ms]

        spawn(fn ->
          :timer.send_interval(every, :tick)
          heartbeat_loop(rt, h, on_report)
        end)
      end

    %{
      handle: h,
      rt: rt,
      stop: fn ->
        if looper, do: Process.exit(looper, :kill)
        Runtime.close(rt)
      end
    }
  end

  defp heartbeat_loop(rt, h, on_report) do
    receive do
      :tick ->
        Runtime.post(rt, h, %{from: "clock", channel: "timer", text: "tick"})

        if Runtime.snapshot(h).state == :idle do
          r =
            Runtime.run_turn(
              rt,
              h,
              "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK."
            )

          if not r.is_error and not String.contains?(r.text, @heartbeat_ok) and is_function(on_report, 1),
            do: on_report.(r.text)
        end

        heartbeat_loop(rt, h, on_report)
    end
  end
end

File: golang/spike/agents.go
// Level-1/2/bridge UX sugar over the substrate (see audit doc "The UX").
// Go port of js/spike/agents.ts. One new noun: Agent(). Everything compiles to
// the six verbs in runtime.go.
//
//	explore := spike.NewAgent("explore", spike.AgentSpec{Does: "...", Tools: []tn.Tool{lookup}})
//	coder := spike.NewAgent("coder", spike.AgentSpec{Does: "...", SoulFile: "AGENTS.md", Team: []*spike.Agent{explore}})
//	r, _ := coder.Run(rtOpts, "fix the bug")
//
//	mia := spike.AgentFromDir("~/mia", nil) // the directory IS the agent
package spike

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	tn "github.com/muthuishere/toolnexus/golang"
)

// AgentSpec declares an agent at the UX layer.
type AgentSpec struct {
	// Does is what the delegating model sees — the routing description.
	Does string
	// Tools is the toolkit VIEW for this agent (scoping = the whole security model).
	Tools []tn.Tool
	// Soul is identity as inline text (AGENTS.md / SOUL.md content).
	Soul string
	// SoulFile is the path form of Soul — read at registry-build time.
	SoulFile string
	// Team: task-tool targets — listing agents here IS the subagent wiring.
	Team []*Agent
	Budget *Budget
	// Model id; "" ⇒ "inherit" (runtime default).
	Model string
	// WaitFor is the §10 interpreter authority for this agent's subtree.
	WaitFor func(tn.Request) (tn.Answer, error)
	OnSpawn func(h *Handle)
	OnClose func(h *Handle, reason string)
}

// Agent is the one new noun.
type Agent struct {
	Name string
	Spec AgentSpec
}

// NewAgent builds an Agent.
func NewAgent(name string, spec AgentSpec) *Agent {
	return &Agent{Name: name, Spec: spec}
}

// Registry collects this agent + its whole team graph into a runtime registry.
func (a *Agent) Registry() map[string]AgentDef {
	acc := map[string]AgentDef{}
	a.registryInto(acc)
	return acc
}

func (a *Agent) registryInto(acc map[string]AgentDef) {
	if _, ok := acc[a.Name]; ok {
		return
	}
	soul := a.Spec.Soul
	if a.Spec.SoulFile != "" {
		if b, err := os.ReadFile(a.Spec.SoulFile); err == nil {
			soul = string(b)
		}
	}
	model := a.Spec.Model
	if model == "" {
		model = "inherit"
	}
	targets := make([]string, 0, len(a.Spec.Team))
	for _, t := range a.Spec.Team {
		targets = append(targets, t.Name)
	}
	acc[a.Name] = AgentDef{
		Name:         a.Name,
		Description:  a.Spec.Does,
		SystemPrompt: soul,
		Model:        model,
		Tools:        a.Spec.Tools,
		Budget:       a.Spec.Budget,
		WaitFor:      a.Spec.WaitFor,
		OnSpawn:      a.Spec.OnSpawn,
		OnClose:      a.Spec.OnClose,
		// team scoping: task targets = ONLY this agent's team (not the whole registry)
		TaskTargets: targets,
	}
	for _, t := range a.Spec.Team {
		t.registryInto(acc)
	}
}

// Run — Level 1: one-shot. Builds a runtime, runs to completion, tears down.
// The runtime is returned alongside for durable resume (JS attaches it to the
// result; Go returns it).
func (a *Agent) Run(rtOpts RuntimeOptions, prompt string) (TaskResult, *AgentRuntime) {
	rtOpts.Registry = a.Registry()
	rt := NewAgentRuntime(rtOpts)
	h, err := rt.Spawn(rt.Root, a.Name, nil)
	if err != nil {
		return TaskResult{Text: err.Error(), IsError: true, Status: "done"}, rt
	}
	rt.Wake(h, prompt)
	r := rt.Wait(h, 0)
	if r.Status != "pending" {
		rt.Close(rt.Root, nil)
	}
	return r, rt
}

// AsTool — the bridge: an Agent IS a Tool — drop it into the OLD API's ExtraTools.
func (a *Agent) AsTool(rtOpts RuntimeOptions) tn.Tool {
	return tn.Tool{
		Name:        a.Name,
		Description: a.Spec.Does,
		InputSchema: tn.JSONSchema{
			"type":       "object",
			"properties": map[string]any{"prompt": map[string]any{"type": "string"}},
			"required":   []any{"prompt"},
		},
		Source: tn.SourceCustom,
		Execute: func(args map[string]any, _ *tn.ToolContext) (tn.ToolResult, error) {
			prompt, _ := args["prompt"].(string)
			r, _ := a.Run(rtOpts, prompt)
			return tn.ToolResult{
				Output:   r.Text,
				IsError:  r.IsError,
				Metadata: map[string]any{"agent": a.Name, "turns": r.Turns, "totalTokens": r.TotalTokens},
			}, nil
		},
	}
}

// BootstrapOrder — Level 2 discovery order (all optional, openclaw order).
var BootstrapOrder = []string{"AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md"}

const maxBootstrapFileBytes = 2 * 1024 * 1024

// AgentFromDir — Level 2: the directory IS the agent. Discovered files are
// injected as named "## <file>" sections at session start (cache doctrine:
// onSpawn reads memory in).
func AgentFromDir(dir string, opts *AgentSpec, name string) *Agent {
	var sections []string
	for _, f := range BootstrapOrder {
		p := filepath.Join(dir, f)
		b, err := os.ReadFile(p)
		if err != nil {
			continue
		}
		body := string(b)
		if len(b) > maxBootstrapFileBytes {
			body = body[:maxBootstrapFileBytes] + "\n[truncated: file exceeds bootstrap cap]"
		}
		sections = append(sections, fmt.Sprintf("## %s\n\n%s", f, strings.TrimSpace(body)))
	}
	if name == "" {
		name = filepath.Base(dir)
	}
	spec := AgentSpec{}
	if opts != nil {
		spec = *opts
	}
	if spec.Does == "" {
		spec.Does = "persona agent from " + dir
	}
	spec.Soul = strings.Join(sections, "\n\n")
	return NewAgent(name, spec)
}

// HeartbeatOK is the silent-wake sentinel — a heartbeat turn answering this
// produces no outbound report.
const HeartbeatOK = "HEARTBEAT_OK"

const heartbeatPrompt = "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK."

// StartedAgent is a long-lived persona started by StartAgent.
type StartedAgent struct {
	Handle *Handle
	RT     *AgentRuntime
	stop   chan struct{}
	ticker *time.Ticker
}

// Stop halts the heartbeat and closes the runtime (leaf-first cascade).
func (s *StartedAgent) Stop() {
	if s.ticker != nil {
		s.ticker.Stop()
	}
	close(s.stop)
	s.RT.Close(s.RT.Root, nil)
}

// StartAgent — Level 2: start a long-lived persona. The heartbeat posts a tick
// to its own inbox (unsolicited rail; ticks coalesce) and wakes it at idle;
// HEARTBEAT_OK results stay silent.
func StartAgent(a *Agent, rtOpts RuntimeOptions, everyMs int, onReport func(text string)) (*StartedAgent, error) {
	rtOpts.Registry = a.Registry()
	rt := NewAgentRuntime(rtOpts)
	handle, err := rt.Spawn(rt.Root, a.Name, nil)
	if err != nil {
		return nil, err
	}
	s := &StartedAgent{Handle: handle, RT: rt, stop: make(chan struct{})}
	if everyMs > 0 {
		s.ticker = time.NewTicker(time.Duration(everyMs) * time.Millisecond)
		go func() {
			for {
				select {
				case <-s.stop:
					return
				case <-s.ticker.C:
					rt.Post(handle, InboxItem{From: "clock", Channel: "timer", Text: "tick"})
					if rt.StateOf(handle) == StateIdle {
						go func() {
							r := rt.RunTurn(handle, heartbeatPrompt)
							if !r.IsError && !strings.Contains(r.Text, HeartbeatOK) && onReport != nil {
								onReport(r.Text)
							}
						}()
					}
				}
			}
		}()
	}
	return s, nil
}
